examples/vae_text/vae_train.py

```diff
@@ -196,111 +193,82 @@ def forward(self, data_batch: Dict, kl_weight: float) -> Dict:
-                (self._config.batch_size,),
-                self._bos_token_id,
-                dtype=torch.long, device=device)
-            end_token = self._eos_token_id
-            if self._config.decoder_type == "lstm":
-                def _cat_embedder(ids):
-                    """Concatenates latent variable to input word embeddings
-                    """
-                    embedding = self.decoder_w_embedder(ids.to(device))
-                    return torch.cat([embedding, latent_z], 1)
-
-                infer_helper = self.decoder.create_helper(
-                    embedding=_cat_embedder,
-                    decoding_strategy='infer_sample',
-                    start_tokens=start_tokens,
-                    end_token=end_token)
+               decoder_states: Union[Tensor, Tuple],
+               helper: Optional[Helper],
+               seq_lengths: Optional[Tensor] = None,
+               max_decoding_length: int = None,
+               text_ids: torch.LongTensor = None) \
+            -> Union[BasicRNNDecoderOutput, TransformerDecoderOutput]:
 
-                outputs, _, _ = self.decoder(
-                    initial_state=dcdr_states,
-                    helper=infer_helper,
-                    max_decoding_length=100)
+        if self._config.decoder_type == "lstm":
+            outputs, _, _ = self.decoder(
+                initial_state=decoder_states,
+                inputs=text_ids,
+                helper=helper,
+                sequence_length=seq_lengths,
+                max_decoding_length=max_decoding_length)
+        else:
+            if text_ids is not None:
+                batch_size = text_ids.size(0)
+                max_seq_len = text_ids.size(1) - 1
+                batch_max_seq_len = text_ids.new_full(
+                    (batch_size,), max_seq_len, dtype=torch.long)
+                output_p_embed = self.decoder_p_embedder(
+                    sequence_length=batch_max_seq_len)
+                output_w_embed = self.decoder_w_embedder(text_ids[:, :-1])
+                output_w_embed = output_w_embed * \
+                    self._config.hidden_size ** 0.5
+                output_embed = output_w_embed + output_p_embed
             else:
-                def _embedding_fn(ids, times):
-                    w_embed = self.decoder_w_embedder(ids)
-                    p_embed = self.decoder_p_embedder(times)
-                    return w_embed * config.hidden_size ** 0.5 + p_embed
-
-                outputs, _ = self.decoder(
-                    memory=dcdr_states,
-                    memory_sequence_length=torch.ones(dcdr_states.size(0)),
-                    max_decoding_length=100,
-                    decoding_strategy='infer_sample',
-                    embedding=_embedding_fn,
-                    start_tokens=start_tokens,
-                    end_token=end_token)
+                output_embed = None
+            outputs = self.decoder(
+                inputs=output_embed,
```

Comment:
And here you would pass `text_ids` to `inputs`. No need to compute `output_embed` here.

Reply:
Got it, after implementing `token_pos_embedder`.